sdk/typescript/src/api/http-transport.ts

```diff
@@ -0,0 +1,90 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+import { RequestManager, HTTPTransport, Client } from '@open-rpc/client-js';
+import { PACKAGE_VERSION, TARGETED_RPC_VERSION } from '../version.js';
+import type { WebsocketClientOptions } from '../rpc/websocket-client.js';
+import { WebsocketClient } from '../rpc/websocket-client.js';
+
+/**
+ * An object defining headers to be passed to the RPC server
+ */
+export type HttpHeaders = { [header: string]: string };
+
+interface SuiHTTPTransportOptions {
+	network: SuiNetwork;
+	rpc?: {
+		headers?: HttpHeaders;
+		url?: string;
+	};
+	websocket?: WebsocketClientOptions & {
+		url?: string;
+	};
+}
+
+export interface SuiTransportRequestOptions {
+	method: string;
+	params: unknown[];
+}
+
+// eslint-disable-next-line @typescript-eslint/ban-types
+export type SuiNetwork =
+	| 'mainnet'
+	| 'testnet'
+	| 'devnet'
+	| 'localnet'
+	| `http${'s' | ''}://${string}`;
+
+const networkToEndpoint: Record<SuiNetwork, string> = {
+	mainnet: 'https://fullnode.mainnet.sui.io:443',
+	testnet: 'https://fullnode.testnet.sui.io:443',
+	devnet: 'https://fullnode.devnet.sui.io:443',
+	localnet: 'http://127.0.0.1:9000',
+};
+
+export interface SuiTransportSubscribeOptions<T> {
+	method: string;
+	unsubscribe: string;
+	params: unknown[];
+	onMessage: (event: T) => void;
+}
+
+export interface SuiTransport {
+	request<T = unknown>(input: SuiTransportRequestOptions): Promise<T>;
+	subscribe<T = unknown>(input: SuiTransportSubscribeOptions<T>): Promise<() => Promise<boolean>>;
+}
+
+export class SuiHTTPTransport implements SuiTransport {
+	private rpcClient: Client;
```

Comment:
I know we want to entirely reimplement the underlying pieces of this. Do we want to at least start this now so that e.g. the errors they throw are consistent between transports?

Reply:
Do you mean standardize the Errors between websockets and RPC, or create a set of standard errors that someone else could reuse if they were implementing their own Transport?

I'm not sure I see a lot of benefits in doing it now, but open to your thoughts here.   Currently, as far as I can tell, the subscription API doesn't actually have a way to surface errors (which isn't great), and the RPC client should throw `JSONRPCError`s.  I think this may make more sense to tackle when when we change the underlying implementation, but I could also see an argument for creating some sort of wrapper around the RPC calls that throws our own error.  I don't see a good way to do it for the subscriptions though (without a large change to how subscriptions work)